enclave/impl/src/message_bus/mod.rs

```diff
@@ -0,0 +1,753 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+//! Object representing trusted storage for message bus contents.
+//! This stores messages by ids, and associates a fixed size storage to each
+//! recipient containing a list of in-flight ids.
+//! The message bus object ensures that these are in sync and that expired
+//! messages are purged lazily over time.
+
+use aligned_cmov::{
+    subtle::{Choice, ConditionallySelectable, ConstantTimeEq, ConstantTimeLess},
+    typenum::{U1008, U1024, U16, U24, U32, U4096, U64, U8, U992},
+    A8Bytes, Aligned, AsAlignedChunks, AsNeSlice, CMov,
+};
+use alloc::boxed::Box;
+use generic_array::sequence::Split;
+use mc_common::logger::Logger;
+use mc_crypto_keys::{RistrettoPublic, RistrettoSignature};
+use mc_crypto_rand::{McRng, RngCore};
+use mc_grapevine_enclave_api::QueryError;
+use mc_grapevine_types::{
+    QueryRequest, QueryResponse, GRAPEVINE_CHALLENGE_SIGNING_CONTEXT, REQUEST_TYPE_CREATE,
+    REQUEST_TYPE_DELETE, REQUEST_TYPE_UPDATE, STATUS_CODE_INTERNAL_ERROR,
+    STATUS_CODE_INVALID_RECIPIENT, STATUS_CODE_SUCCESS,
+};
+use mc_oblivious_map::CuckooHashTableCreator;
+use mc_oblivious_ram::PathORAM4096Z4Creator;
+use mc_oblivious_traits::{
+    OMapCreator, ORAMStorageCreator, ObliviousHashMap, /* OMAP_FOUND, */ OMAP_INVALID_KEY,
+    OMAP_NOT_FOUND, OMAP_OVERFLOW,
+};
+
+mod marshalled_record;
+use marshalled_record::MarshalledRecord;
+
+// internal constants
+
+// KeySize and ValueSize reflect the needs of message bus
+// We must choose an oblivious map algorithm that can support that
+// The message bus key is one 16 byte id
+// The message bus value is 1008 remaining bytes. These get mapped into protobuf
+// as described in the README, and the user's payload is somewhere in there.
+type MessageBusKeySize = U16;
+type MessageBusValueSize = U1008; // 1008 = 1024 - 16
+
+// Type alias for the key to the message bus map (a "Message Id")
+type MsgId = A8Bytes<MessageBusKeySize>;
+
+// The recipient-to-message-id map maps recipient pubkeys to the ids of their
+// messages. A recipient key is 32 bytes.
+// The value is 1024 - 32 bytes, divided into 24 byte chunks which are an id and
+// a timestamp.
+type RecipientKeySize = U32;
+type RecipientValueSize = U992; // 992 = 1024 - 32
```

Comment:
So test test my understanding, this means up to 41 messages can be queued up for recipient?